File: js/packages/teams-ai/src/Messages.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    ActivityTypes,
    Channels,
    INVOKE_RESPONSE_KEY,
    InvokeResponse,
    TaskModuleResponse,
    TaskModuleTaskInfo,
    TurnContext
} from 'botbuilder';
import { Application } from './Application';
import { TurnState } from './TurnState';

export enum MessageInvokeNames {
    FETCH_INVOKE_NAME = `message/fetchTask`
}

/**
 * TaskModules class to enable fluent style registration of handlers related to Task Modules.
 * @template TState Type of the turn state object being persisted.
 */
export class Messages<TState extends TurnState> {
    private readonly _app: Application<TState>;

    /**
     * Creates a new instance of the TaskModules class.
     * @param {Application} app Top level application class to register handlers with.
     */
    public constructor(app: Application<TState>) {
        this._app = app;
    }

    /**
     * Registers a handler to process the initial fetch of the task module.
     * @remarks
     * Handlers should respond with either an initial TaskInfo object or a string containing
     * a message to display to the user.
     * @template TData Optional. Type of the data object being passed to the handler.
     * @param {(context: TurnContext, state: TState, data: TData) => Promise<TaskModuleTaskInfo | string>} handler - Function to call when the handler is triggered.
     * @param {TurnContext} handler.context - Context for the current turn of conversation with the user.
     * @param {TState} handler.state - Current state of the turn.
     * @param {TData} handler.data - Data object passed to the handler.
     * @returns {Application<TState>} The application for chaining purposes.
     */
    public fetch<TData extends Record<string, any> = Record<string, any>>(
        handler: (context: TurnContext, state: TState, data: TData) => Promise<TaskModuleTaskInfo | string>
    ): Application<TState> {
        this._app.addRoute(
            async (context) => {
                return (
                    context?.activity?.type === ActivityTypes.Invoke &&
                    context?.activity?.name === MessageInvokeNames.FETCH_INVOKE_NAME
                );
            },
            async (context, state) => {
                if (context?.activity?.channelId === Channels.Msteams) {
                    const result = await handler(context, state, context.activity.value?.data ?? {});

                    if (!context.turnState.get(INVOKE_RESPONSE_KEY)) {
                        // Format invoke response
                        let response: TaskModuleResponse;
                        if (typeof result == 'string') {
                            // Return message
                            response = {
                                task: {
                                    type: 'message',
                                    value: result
                                }
                            };
                        } else {
                            // Return card
                            response = {
                                task: {
                                    type: 'continue',
                                    value: result
                                }
                            };
                        }

                        // Queue up invoke response
                        await context.sendActivity({
                            value: { body: response, status: 200 } as InvokeResponse,
                            type: ActivityTypes.InvokeResponse
                        });
                    }
                }
            },
            true
        );

        return this._app;
    }
}

File: python/packages/ai/teams/messages/messages.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import Awaitable, Callable, Generic, List, TypeVar, Union, cast

from botbuilder.core import InvokeResponse, TurnContext
from botbuilder.core.serializer_helper import serializer_helper
from botbuilder.schema import Activity, ActivityTypes
from botbuilder.schema.teams import (
    TaskModuleContinueResponse,
    TaskModuleMessageResponse,
    TaskModuleResponse,
    TaskModuleTaskInfo,
)

from teams.route import Route
from teams.state import TurnState

FETCH_INVOKE_NAME = "message/fetchTask"

StateT = TypeVar("StateT", bound=TurnState)


class Messages(Generic[StateT]):
    _routes: List[Route[StateT]] = []

    def __init__(self, routes: List[Route[StateT]]) -> None:
        self._routes = routes

    def fetch_task(self) -> Callable[
        [Callable[[TurnContext, StateT, dict], Awaitable[Union[TaskModuleTaskInfo, str]]]],
        Callable[[TurnContext, StateT, dict], Awaitable[Union[TaskModuleTaskInfo, str]]],
    ]:
        """
        Adds a route for handling the message fetch task activity.
        This method can be used as either a decorator or a method.

        ```python
        # Use this method as a decorator
        @app.messages.fetch_task()
        async def fetch_task(context: TurnContext, state: TurnState, data: dict):
            print(f"Execute with data: {data}")
            return True

        # Pass a function to this method
        app.messages.fetch_task()(fetch_task)
        ```
        """

        # Create route selector for the handler
        def __selector__(context: TurnContext) -> bool:
            return (
                context.activity.type == ActivityTypes.invoke
                and context.activity.name == FETCH_INVOKE_NAME
            )

        def __call__(
            func: Callable[
                [TurnContext, StateT, dict],
                Awaitable[Union[TaskModuleTaskInfo, str]],
            ],
        ) -> Callable[
            [TurnContext, StateT, dict],
            Awaitable[Union[TaskModuleTaskInfo, str]],
        ]:
            async def __invoke__(context: TurnContext, state: StateT):
                res = await func(context, state, cast(dict, context.activity.value))
                await self._invoke_task_response(context, res)
                return True

            self._routes.append(Route[StateT](__selector__, __invoke__, True))
            return func

        return __call__

    async def _invoke_task_response(
        self, context: TurnContext, body: Union[TaskModuleTaskInfo, str]
    ):
        if context._INVOKE_RESPONSE_KEY in context.turn_state:
            return

        response = TaskModuleResponse(task=TaskModuleContinueResponse(value=body))

        if isinstance(body, str):
            response = TaskModuleResponse(task=TaskModuleMessageResponse(value=body))

        await context.send_activity(
            Activity(
                type=ActivityTypes.invoke_response,
                value=InvokeResponse(body=serializer_helper(response), status=200),
            )
        )
